### PR TITLE
Add Frontend Validation for Integers on Search Page

### LINKED ISSUE #13326: No quotes when field type is int
https://demo.phpmyadmin.net/QA_4_9/tbl_select.php?db=sakila&table=actor

    SELECT * FROM `actor` WHERE `actor_id` = test LIMIT 0, 25 :(

I think it should put quotes around test..

### LINKED ISSUE #15782: Allow only Integer values in INT field
### Is your feature request related to a problem? Please describe.

We can enter non-integer values in Integer fields, Which is not required/should not be allowed.

### Describe the solution you'd like

We can allow only integer values to be entered in the field, by adding front end check like type- number.

I would like to work on this.

image

### PR DESCRIPTION
**Pull request by thesmallstar**
### Description

**Front End validation is added on the Insert and Search Pages**

image

image

**I am not sure about updating the Number field width for the entire theme, by default it was too small,any suggestions?**

Fixes #13326  

Before submitting pull request, please review the following checklist:

- [ ] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [ ] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [ ] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [ ] Every commit has a descriptive commit message.
- [ ] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [ ] Any new functionality is covered by tests.

### CODE REVIEW

**Comment by williamdes** [comment]
@thesmallstar could you use a validator because not all browsers support the number input ?

**Comment by thesmallstar** [comment]
@williamdes, a validator is already added validator is activated when the user enters the value. Here we are trying to stop the user from entering a value that is not a number. 

**Browser Support:**

All major browser including IE(internet explorer) support type = number since 2012, the last to include type=number was opera in 2014. As seen in in: [here](https://caniuse.com/#feat=input-number)

image

**Any suggestions on how should I proceed on this?**

**Comment by williamdes** [comment]
Okay then
But If possible to use the validator that is currently implemented?

File: libraries/classes/Controllers/Table/SearchController.php [hunk 1]
```diff
@@ -953,6 +953,11 @@ public function getColumnProperties($search_index, $column_index)
             '',
             ''
         );
+        //true_type is used to set type as number for INT
+        $true_type  = preg_replace('@\(.*@s', '', $type);
+        if (in_array((string)$true_type, $this->dbi->types->getIntegerTypes())){
```
Comment (williamdes) [inline libraries/classes/Controllers/Table/SearchController.php]: I am just not quite fan of this piece of code, can it be improved ?

**Comment by thesmallstar** [comment]
@williamdes, can you help me with this?
Actually there is a working validator which does this:
image

Can we implement a validator that does the same work as type=number, that does not let the user type any non-int values?

File: libraries/classes/Controllers/Table/SearchController.php (hunk 1 above)
Comment (thesmallstar) [reply libraries/classes/Controllers/Table/SearchController.php]: sure, working on this.

**Comment by williamdes** [comment]
"Actually there is a working validator which does this:" :muscle: 
Perfect, this is exactly what I wanted :100: 
Can you push the code ?

**Comment by thesmallstar** [comment]
@williamdes  actually that is **already** the part of PHPMyAdmin current QA_5_0 branch  :P :P
I am trying to implement/the issue suggested was: Do not let the user type the non-int values in the INT field. 
**The validator is letting the user type and then it verifies. 
I am trying that the user cannot type it in the first place.**

**Comment by williamdes** [comment]
"The validator is letting the user type and then it verifies." okay, please check on "Search" page like the user

"I am trying that the user cannot type it in the first place." this is an enhancement and must be done in another pr ;)

So for me what must be done is: add the validator on search page

image


image

**Comment by thesmallstar** [comment]
@williamdes okay, so I am implementing the validation on the Search page.
**In this PR.**

**Also, I will be making another PR for implementing the enhancement(In master Branch).**
**on it. :D :)**

**Comment by williamdes** [comment]
@thesmallstar are you still working on this issue ?

**Comment by thesmallstar** [comment]
Yes, I had some college events going so I was busy, is it urgent? I can do it in a day

**Comment by williamdes** [comment]
> Yes, I had some college events going so I was busy, is it urgent? I can do it in a day

No problem, not very urgent for now ;)

**Comment by thesmallstar** [comment]
@williamdes  can you help me on how to proceed with adding a validator?  I looked around, but  I am missing pieces on how to implement this with Sanitize.

**Comment by williamdes** [comment]
@thesmallstar The input needs to have `data-type="INT"` and more code enabled to work
Here is my unfinished diff, you can use it:
```diff
diff --git a/libraries/classes/Controllers/Table/SearchController.php b/libraries/classes/Controllers/Table/SearchController.php
index 3f2ceae35c..57026487fb 100644
--- a/libraries/classes/Controllers/Table/SearchController.php
+++ b/libraries/classes/Controllers/Table/SearchController.php
@@ -12,6 +12,7 @@ namespace PhpMyAdmin\Controllers\Table;
 use PhpMyAdmin\DatabaseInterface;
 use PhpMyAdmin\Relation;
 use PhpMyAdmin\Response;
+use PhpMyAdmin\Sanitize;
 use PhpMyAdmin\Sql;
 use PhpMyAdmin\Template;
 use PhpMyAdmin\Util;
@@ -935,9 +936,10 @@ class SearchController extends AbstractController
         //Gets column's type and collation
         $type = $this->_columnTypes[$column_index];
         $collation = $this->_columnCollations[$column_index];
+        $cleanType = preg_replace('@\(.*@s', '', $this->_columnTypes[$column_index]);
         //Gets column's comparison operators depending on column type
         $typeOperators = $this->dbi->types->getTypeOperatorsHtml(
-            preg_replace('@\(.*@s', '', $this->_columnTypes[$column_index]),
+            $cleanType,
             $this->_columnNullFlags[$column_index],
             $selected_operator
         );
@@ -953,8 +955,28 @@ class SearchController extends AbstractController
             '',
             ''
         );
+        $html_attributes = '';
+        if (in_array($cleanType, $this->dbi->types->getIntegerTypes())) {
+            $extracted_columnspec = Util::extractColumnSpec(
+                $type
+            );
+            $is_unsigned = $extracted_columnspec['unsigned'];
+            $min_max_values = $this->dbi->types->getIntegerRange(
+                $cleanType,
+                ! $is_unsigned
+            );
+            $html_attributes = 'min="' . $min_max_values[0] . '" '
+                . 'max="' . $min_max_values[1] . '"';
+            $type = 'INT';
+        }
+        // Example: onchange="return verificationsAfterFieldChange('b39fef6a9c6e9bbf9d7f14ab2fd7e484', '0','smallint(5) unsigned')"
+        $html_attributes = 'onchange'
+        . "=\"return verificationsAfterFieldChange('"
+        . Sanitize::escapeJsString($column['Field_md5']) . "', '"
+        . Sanitize::escapeJsString($jsvkey) . "','" . $type . "')\"";
         $value = $this->template->render('table/search/input_box', [
             'str' => '',
+            'html_attributes' => $html_attributes,
             'column_type' => (string) $type,
             'column_id' => 'fieldID_',
             'in_zoom_search_edit' => false,
diff --git a/templates/table/search/input_box.twig b/templates/table/search/input_box.twig
index 5c30535bd7..9cdf7ae5e6 100644
--- a/templates/table/search/input_box.twig
+++ b/templates/table/search/input_box.twig
@@ -87,6 +87,9 @@
         {% set the_class = the_class ~ ' bit' %}
     {% endif %}
     <input type="text"
+        data-type="{{ column_type }}"
+        {# My code needs to be checked #}
+        {{ html_attributes }}
         name="criteriaValues[{{ column_index }}]"
         size="40"
         class="{{ the_class }}"
```


I hope this helps you.
You will need to fix the front-end part in order to trigger the validator code
Maybe also split the validator function in two parts.
The diff does not work but is near to work

**Comment by thesmallstar** [comment]
@williamdes  
1. I am writing a different onChange function for Search page(Is this okay?, all fields are not common).
2. Can you tell me for what exactly the name is converted to md5 hash?

**Comment by williamdes** [comment]
Hi @thesmallstar 

> I am writing a different onChange function for Search page(Is this okay?, all fields are not common).

I it is okay If you have very minimal code duplication

> Can you tell me for what exactly the name is converted to md5 hash?

To have a non breaking html "id", special chars in column names make it impossible to use the name directly

**Comment by williamdes** [comment]
Duplicate of #15846

**Pull request closed without merging**